rollback on panic

ashie1287 opened the pull request:
Fixes #76 

~I tried to do the minimum amount of changes to fix the issue.~
UPDATE: did a very small amount of clean up to help me understand a bit better.

There are currently no tests around this. I explored the current tests, and after a while (lots of abstraction), I was able to understand them for the most part. However, it seems that currently (and please correct me if I'm wrong) there are no existing tests actually testing the existing `defer` statement? I would have expected at least some testing of a "non-retryable" error resulting in a rollback, but either there are not any or I couldn't find them.

If someone could advise me on what tests for rollback (whether or not via `panic` or "non-retryable" error, probably should add tests for both) should look like, I'd much appreciate it and am happy to work on them.

Let me know if you have questions, thank you! 🙏

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach-go/119)
<!-- Reviewable:end -->

ashie1287 commented on go.mod:
```diff
@@ -3,7 +3,7 @@ module github.com/cockroachdb/cockroach-go/v2
 
 require (
-	github.com/gofrs/flock v0.8.1 // indirect
+	github.com/gofrs/flock v0.8.1
```
Just from `go mod download; go mod tidy` on my machine, happy to revert if I got it wrong.

rafiss commented on crdb/common.go:
```diff
@@ -58,31 +66,35 @@ func ExecuteInTx(ctx context.Context, tx Tx, fn func() error) (err error) {
 		}
-		// We got an error; let's see if it's a retryable one and, if so, restart.
+
+		// We got an error (i.e. err != nil) if at this point
```
nit: this comment needs to be fixed

ashie1287 replied:
Not 100% sure what you mean, but put it back to exactly what was there. 👍

rafiss reviewed (changes requested):
thanks for this change! it looks good overall -- can you adjust the comment and rebase this?

ashie1287 commented:
@rafiss Thanks for taking a look.

Rebased against `main`, fixed the comment, and actually reverted a few changes, so the change set is now smaller than before (had a misunderstanding of Go named return values).

Just to re-iterate: still no testing around these changes, please advise if you'd like me to add some and if you have any ideas/thoughts on how to accomplish that.

Thanks again! 🙏

rafiss reviewed (approved):
this lgtm! i wouldn't worry too much about testing this

rafiss commented:
If you do want to keep working on the tests for this, here's what you can try:
- implement your own (testing only) version of the `crdb.Tx` interface that can detect when Commit and Rollback have been called.
- call `crdb.ExecuteInTx` with your custom `Tx` and a function that panics. (and another test for a function that returns an error.
- make sure Commit has not been called, and that Rollback has been called.
